frontend/src/services/validate.service.ts

```diff
@@ -0,0 +1,137 @@
+import Papa from 'papaparse'
+import { keys, difference } from 'lodash'
+
+export interface CsvStatus {
+  csvFilename?: string
+  csvError?: string
+  numRecipients?: number
+  preview?: string
+}
+
+/**
+ * Ensure that the keywords found in the template is all in the csv headers and each row.
+ * Also verifies that the the headers and each row has the following keys:
+ * recipient,password
+ *
+ */
+export async function validateCsv(
+  file: File,
+  templateParams: Array<string>
+): Promise<CsvStatus> {
+  try {
+    await validateHeaders(file, templateParams)
+    const numRows = await validateRows(file, templateParams)
+    return Promise.resolve({
+      csvError: '',
+      numRecipients: numRows,
+      // TODO: Change this to an actual preview
+      preview: 'This is a preview of message B',
+      csvFilename: file.name,
+    })
+  } catch (err) {
+    // if there is an error, return a csv status with the error message
+    return Promise.resolve({
+      csvError: err,
+      numRecipients: 0,
+      preview: 'This is a preview of message B',
+      csvFilename: file.name,
+    })
+  }
+}
+
+/**
+ * Read only the headers of the csv file
+ *
+ */
+async function validateHeaders(
+  file: File,
+  templateParams: Array<string>
+): Promise<void> {
+  await new Promise((resolve, reject) => {
+    Papa.parse(file, {
+      header: false,
+      delimiter: ',',
+      step: function (_, parser: Papa.Parser) {
+        // Checks header only
+        parser.pause()
+        parser.abort()
+      },
+      complete: function (results) {
+        const headers = results.data as Array<string>
+        try {
+          checkTemplateKeys(headers, templateParams)
+          checkEssentialKeys(headers)
+        } catch (e) {
+          reject(e.message)
+        }
+        resolve()
+      },
+    })
+  })
+}
+
+/**
+ * Read the file row by row and ensure that all the required params are there.
+ * Returns the number of rows in the csv file
+ */
+async function validateRows(
+  file: File,
+  templateParams: Array<string>
+): Promise<number> {
+  return await new Promise((resolve, reject) => {
+    let count = 0
+    Papa.parse(file, {
+      header: true,
+      delimiter: ',',
+      skipEmptyLines: true,
+      step: function (step) {
+        // If there is errors, append to the errors array
+        if (step.errors.length != 0) {
+          reject(`Error parsing file.`)
+        }
+        // Check params
+        const params = step.data
+
+        try {
+          const paramKeys = keys(params)
+          checkTemplateKeys(paramKeys, templateParams)
```

Comment:
Ideally i would like to template it instead of just checking if the params contain the keys. I am also kinda worried about the performance, since for each row we are doing `keys(params) and then finding out the difference between each array.